src/SearchImpl/Indexing/CentralizedIndexingActivityQueue.cs

```diff
@@ -0,0 +1,291 @@
+        /// Releases the finished activities and executes the loaded ones in asynchron way.
+        /// </summary>
+        private static int ExecuteActivities(IndexingActivityBase waitingActivity, bool systemStart)
+        {
+            int[] waitingActivityIds;
+            lock (_waitingActivitiesSync)
+            {
+                if (waitingActivity != null)
+                {
+                    if (_waitingActivities.TryGetValue(waitingActivity.Id, out IndexingActivityBase olderWaitingActivity))
+                    {
+                        // if exists, attach the current to existing.
+                        olderWaitingActivity.Attach(waitingActivity);
+                        // do not load executables because wait-polling cycle is active.
+                        return 0;
+                    }
+                    // add to waiting list
+                    _waitingActivities.Add(waitingActivity.Id, waitingActivity);
+                }
+                // get id array of waiting activities
+                waitingActivityIds = _waitingActivities.Keys.ToArray();
+            }
+
+            // load some executable activities and currently finished ones
+            var loadedActivities = DataProvider.Current.LoadExecutableIndexingActivities(
+                IndexingActivityFactory.Instance,
+                MaxCount,
+                RunningTimeoutInSeconds,
+                waitingActivityIds,
+                out int[] finishedActivitiyIds);
+            SnTrace.IndexQueue.Write("CIAQ: loaded: {0}, waiting: {1}, finished: {2}, tasks: {3}", loadedActivities.Length, _waitingActivities.Count, finishedActivitiyIds.Length, _activeTasks);
+
+            // release finished activities
+            if (finishedActivitiyIds.Length > 0)
+            {
+                lock (_waitingActivitiesSync)
+                {
+                    foreach (var finishedActivitiyId in finishedActivitiyIds)
+                    {
+                        if (_waitingActivities.TryGetValue(finishedActivitiyId, out IndexingActivityBase finishedActivity))
+                        {
+                            _waitingActivities.Remove(finishedActivitiyId);
+                            finishedActivity.Finish();
+                        }
+                    }
+                }
+            }
+
+            // execute loaded activities
+            foreach (var loadedActivity in loadedActivities)
+            {
+                if (systemStart)
+                    loadedActivity.IsUnprocessedActivity = true;
+
+                if (waitingActivity == null)
+                    // polling branch
+                    Task.Run(() => Execute(loadedActivity));
+                else
+                    // UI branch: pay attention to executing waiting instance, because that is needed to be released (loaded one can be dropped).
+                    Task.Run(() => Execute(loadedActivity.Id == waitingActivity.Id ? waitingActivity : loadedActivity));
```

Comment:
If it chooses the waitingActivity over loadedActivity to execute, because it has the same Id, what will become to the loadedActivity (it will never be executed)?

Reply:
Loaded and waiting activity are logically same but different instance and we MUST execute the "waitingActivity".

Reply:
The reason is that the waiting activity .Net object is already used and referenced in other places, so the loaded activity object (with the same id) should not be used as it would be a duplicate.